server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java

```diff
@@ -0,0 +1,167 @@
+ * significantly more than n / m shards. This allocation decider enforces this principle.
+ * This allocation decider excludes any nodes flagged for shutdown from consideration
+ * when computing optimal shard distributions.
+ */
+public class IndexBalanceAllocationDecider extends AllocationDecider {
+
+    private static final Logger logger = LogManager.getLogger(IndexBalanceAllocationDecider.class);
+    private static final String EMPTY = "";
+
+    public static final String NAME = "index_balance";
+
+    private final IndexBalanceConstraintSettings indexBalanceConstraintSettings;
+    private final boolean isStateless;
+
+    private volatile DiscoveryNodeFilters clusterRequireFilters;
+    private volatile DiscoveryNodeFilters clusterIncludeFilters;
+    private volatile DiscoveryNodeFilters clusterExcludeFilters;
+
+    public IndexBalanceAllocationDecider(Settings settings, ClusterSettings clusterSettings) {
+        this.indexBalanceConstraintSettings = new IndexBalanceConstraintSettings(clusterSettings);
+        setClusterRequireFilters(CLUSTER_ROUTING_REQUIRE_GROUP_SETTING.getAsMap(settings));
+        setClusterExcludeFilters(CLUSTER_ROUTING_EXCLUDE_GROUP_SETTING.getAsMap(settings));
+        setClusterIncludeFilters(CLUSTER_ROUTING_INCLUDE_GROUP_SETTING.getAsMap(settings));
+        clusterSettings.addAffixMapUpdateConsumer(CLUSTER_ROUTING_REQUIRE_GROUP_SETTING, this::setClusterRequireFilters, (a, b) -> {});
+        clusterSettings.addAffixMapUpdateConsumer(CLUSTER_ROUTING_EXCLUDE_GROUP_SETTING, this::setClusterExcludeFilters, (a, b) -> {});
+        clusterSettings.addAffixMapUpdateConsumer(CLUSTER_ROUTING_INCLUDE_GROUP_SETTING, this::setClusterIncludeFilters, (a, b) -> {});
+        isStateless = DiscoveryNode.isStateless(settings);
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (indexBalanceConstraintSettings.isDeciderEnabled() == false || isStateless == false || hasNoFilters() == false) {
+            return Decision.single(Decision.Type.YES, NAME, "Decider is disabled.");
+        }
+
+        Index index = shardRouting.index();
+        if (node.hasIndex(index) == false) {
+            return Decision.single(Decision.Type.YES, NAME, "Node does not currently host this index.");
+        }
+
+        assert node.node() != null;
+        assert node.node().getRoles().contains(INDEX_ROLE) || node.node().getRoles().contains(SEARCH_ROLE);
+
+        if (node.node().getRoles().contains(INDEX_ROLE) && shardRouting.primary() == false) {
+            return Decision.single(Decision.Type.YES, NAME, "An index node cannot own search shards. Decider inactive.");
+        }
+
+        if (node.node().getRoles().contains(SEARCH_ROLE) && shardRouting.primary()) {
+            return Decision.single(Decision.Type.YES, NAME, "A search node cannot own primary shards. Decider inactive.");
+        }
```

Comment:
Are we duplicating logic here? this should be covered by `StatelessAllocationDecider`. Perhaps this is a substantial optimisation, in which case I guess it's OK, but if not I'd prefer to keep the deciders focused. 

If it is just here as an optimisation perhaps we could combine into a single check like `co.elastic.elasticsearch.stateless.allocation.StatelessAllocationDecider#canAllocateShardToNode` so it doesn't distract from the focus of this decider?